libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -266,6 +266,21 @@ void ContractLevelChecker::checkAbstractFunctions(ContractDefinition const& _con
 				_contract.annotation().unimplementedFunctions.push_back(function);
 				break;
 			}
+
+	if (
+		_contract.abstract() && _contract.contractKind() != ContractDefinition::ContractKind::Contract
+	)
+		m_errorReporter.typeError(_contract.location(), "Only contracts can be abstract.");
+
+	if (
+		_contract.contractKind() == ContractDefinition::ContractKind::Contract &&
+		!_contract.annotation().unimplementedFunctions.empty() &&
+		!_contract.abstract()
+	)
+		m_errorReporter.typeError(
```

Comment:
@chriseth what about something like this
```
		std::stringstream stream;
		if ( _contract.annotation().unimplementedFunctions.size() == 1)
			stream << "function";
		else
			stream << "functions";
		for (const auto& function : _contract.annotation().unimplementedFunctions) {
			stream << " " << function->name();
			if (function != *(_contract.annotation().unimplementedFunctions.rbegin())) {
				stream << ", ";
			}
		}
		if ( _contract.annotation().unimplementedFunctions.size() == 1)
			stream << " was not implemented.";
		else
			stream << " where not implemented.";
		m_errorReporter.typeError(_contract.location(),
		                          "Contract \"" + _contract.annotation().canonicalName
		                              + "\" should be marked as abstract, because the " + stream.str());
```

Reply:
Let's do this as a follow-up PR.

Reply:
Also, I think the unimplemented functions should be used together with the secondary source location feature.